### PR TITLE
PIM-9582: Merge 4.0 into master

### PR DESCRIPTION
**Pull request by anaisbaunelemaire**
**Description (for Contributor and Core Developer)**

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Todo
| Added legacy Behats               | Todo
| Added acceptance tests            | Todo
| Added integration tests           | Todo
| Changelog updated                 | Todo
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Akeneo/Pim/Enrichment/Bundle/Command/CleanRemovedAttributesFromProductAndProductModelCommand.php [hunk 1]
```diff
@@ -35,12 +37,16 @@ class CleanRemovedAttributesFromProductAndProductModelCommand extends Command
     private int $productBatchSize;
     private ?CleanValuesOfRemovedAttributesInterface $cleanValuesOfRemovedAttributes;
 
+    /** @var EventDispatcher|null */
+    private $eventDispatcher;
```
Comment (ValentinMumble) [inline src/Akeneo/Pim/Enrichment/Bundle/Command/CleanRemovedAttributesFromProductAndProductModelCommand.php:41]:
```suggestion
    private  EventDispatcher $eventDispatcher;
```

File: src/Akeneo/Pim/Enrichment/Bundle/Command/CleanRemovedAttributesFromProductAndProductModelCommand.php [hunk 2]
```diff
@@ -35,12 +37,16 @@ class CleanRemovedAttributesFromProductAndProductModelCommand extends Command
...
+
     public function __construct(
         EntityManagerClearerInterface $entityManagerClearer,
         ProductQueryBuilderFactoryInterface $productQueryBuilderFactory,
         string $kernelRootDir,
         int $productBatchSize,
-        CleanValuesOfRemovedAttributesInterface $cleanValuesOfRemovedAttributes
+        CleanValuesOfRemovedAttributesInterface $cleanValuesOfRemovedAttributes = null,
```
Comment (ValentinMumble) [inline src/Akeneo/Pim/Enrichment/Bundle/Command/CleanRemovedAttributesFromProductAndProductModelCommand.php:47]:
```suggestion
        CleanValuesOfRemovedAttributesInterface $cleanValuesOfRemovedAttributes,
```

File: src/Akeneo/Pim/Structure/Bundle/EventListener/RemoveNonExistingProductValuesSubscriber.php [hunk 3]
```diff
@@ -31,16 +33,21 @@ final class RemoveNonExistingProductValuesSubscriber implements EventSubscriberI
     /** @var string */
     private $jobName;
 
+    /** @var CreateJobInstanceInterface|null */
+    public $createJobInstance;
```
Comment (ValentinMumble) [inline src/Akeneo/Pim/Structure/Bundle/EventListener/RemoveNonExistingProductValuesSubscriber.php:37]:
```suggestion
    public CreateJobInstanceInterface $createJobInstance;
```

File: src/Akeneo/Pim/Structure/Bundle/EventListener/RemoveNonExistingProductValuesSubscriber.php [hunk 4]
```diff
@@ -31,16 +33,21 @@ final class RemoveNonExistingProductValuesSubscriber implements EventSubscriberI
...
+
     public function __construct(
         TokenStorageInterface $tokenStorage,
         JobInstanceRepository $jobInstanceRepository,
         JobLauncherInterface $jobLauncher,
-        string $jobName
+        string $jobName,
+        CreateJobInstanceInterface $createJobInstance = null // @todo remove nullable on master
```
Comment (ValentinMumble) [inline src/Akeneo/Pim/Structure/Bundle/EventListener/RemoveNonExistingProductValuesSubscriber.php:44]:
```suggestion
        CreateJobInstanceInterface $createJobInstance
```

**Review by ValentinMumble — APPROVED** [review]

File: src/Akeneo/Pim/Enrichment/Bundle/Command/CleanRemovedAttributesFromProductAndProductModelCommand.php [hunk 5]
```diff
@@ -35,12 +37,15 @@ class CleanRemovedAttributesFromProductAndProductModelCommand extends Command
...
+    private EventDispatcher $eventDispatcher;
+
     public function __construct(
         EntityManagerClearerInterface $entityManagerClearer,
         ProductQueryBuilderFactoryInterface $productQueryBuilderFactory,
         string $kernelRootDir,
         int $productBatchSize,
-        CleanValuesOfRemovedAttributesInterface $cleanValuesOfRemovedAttributes
+        CleanValuesOfRemovedAttributesInterface $cleanValuesOfRemovedAttributes,
+        EventDispatcher $eventDispatcher
```
Comment (nmarniesse) [inline src/Akeneo/Pim/Enrichment/Bundle/Command/CleanRemovedAttributesFromProductAndProductModelCommand.php:48]: I guess there is a BC break to add in changelog

**Review by nmarniesse — APPROVED** [review]

File: src/Akeneo/Pim/Enrichment/Bundle/Command/CleanRemovedAttributesFromProductAndProductModelCommand.php (hunk 5 above)
Comment (anaisbaunelemaire) [reply src/Akeneo/Pim/Enrichment/Bundle/Command/CleanRemovedAttributesFromProductAndProductModelCommand.php:48]: Thanks for the suggestion @nmarniesse. I just added it in CE.

**Pull request merged**